FixMath.h

```diff
@@ -0,0 +1,1047 @@
+    - all operations between fixed point number is safe (it won't overflow) and preserve the precision. In particular:
+    - only addition, subtraction and multiplication are implemented
+    - any operation between a signed and an unsigned leads to a signed number
+    - resulting numbers will be casted to a type big enough to store the expected values. It follows that it is worth starting with types that are as small as possible to hold the initial value.
+    - all operations between a fixed point number and a native type (int, float, uint) are *not* safe. If the resulting value cannot be represented in the fixed point type it will overflow. Only addition, subtraction, multiplication and right/left shift are implemented.
+    - safe right/left shifts, which return the correct value in the correct type are implemented as .sR<shift>() and .sL<shift>() respectively, shift being the shifting amount. These shifts are basically on
+
+    More specifically on the returned types of the operations between fixed point math types:
+    - Additions:
+    - UFixMath<NI,NF> + UFixMath<_NI,_NF> returns UFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> + anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> + anything_else returns SFixMath<NI,NF>
+    - Subtractions:
+    - UFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI),MAX(NF,_NF)>
+    - SFixMath<NI,NF> - SFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - SFixMath<NI,NF> - UFixMath<_NI,_NF> returns SFixMath<MAX(NI,_NI)+1,MAX(NF,_NF)>
+    - UFixMath<NI,NF> - anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> - anything_else returns SFixMath<NI,NF>
+    - (-)SFixMath<NI,NF> return SFixMath<NI,NF>
+    - (-)UFixMath<NI,NF> return SFixMath<NI,NF>
+    - Multiplications:
+    - UFixMath<NI,NF> * UFixMath<_NI,_NF> returns UFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - SFixMath<NI,NF> * SFixMath<_NI,_NF> returns SFixMath<NI+_NI,NF+_NF>
+    - UFixMath<NI,NF> * anything_else returns UFixMath<NI,NF>
+    - SFixMath<NI,NF> * anything_else returns SFixMath<NI,NF>
+    - Shifts:
+    - UFixMath<NI,NF> .sR<NS> returns UFixMath<NI-NS,NF+NS>
+    - UFixMath<NI,NF> .sL<NS> returns UFixMath<NI+NS,NF-NS>
+    - same for SFixMath.
+      
+   
+*/
+
+
+
+
+#ifndef FIXMATH2_H_
+#define FIXMATH2_H_
+
+#include "IntegerType.h"
+
+#define MOZZI_SHIFTR(x,bits) (bits > 0 ? (x >> (bits)) : (x << (-bits))) // shift right for positive shift numbers, and left for negative ones.
+#define MAX(N1,N2) (N1 > N2 ? N1 : N2)
+#define UBITSTOBYTES(N) (((N-1)>>3)+1)
+#define SBITSTOBYTES(N) (((N)>>3)+1)
```

Comment:
I''m a bit uncertain, whether these should also all get a ```MOZZI_``` prefix. I somewhat lean towards that for safety, even if they are certainly much less likely to clash with anything.

Reply:
(And - quite idly, please don't take this as a request - I'm wondering, whether those could be turned into ```constexpr```s (which would allow placing them in a namespace of something))

Reply:
D'uh, you're undef'ing them at the end of the file, already. Parentheses in MAX will still be a good idea, though, to avoid future surprises.

Reply:
> - I'm wondering, whether those could be turned into constexprs

Indeed, I thought about it. I remember trying and got deceive by some C++17 requirement but I might have done something wrong. Might try again, it would indeed be more C++ like…